platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/style/GeoJsonSourceTests.java

```diff
@@ -41,60 +41,69 @@ protected Class getActivityClass() {
+      GeoJsonSource source = null;
+      try {
+        source = new GeoJsonSource("source", FeatureCollection
+          .fromJson(ResourceUtils.readRawResource(rule.getActivity(), R.raw.test_feature_collection)));
+      } catch (IOException exception) {
+        Timber.e(exception);
       }
+      mapboxMap.addSource(source);
+      mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
     });
   }
 
   @Test
   public void testPointGeometry() {
     validateTestSetup();
-    onView(withId(R.id.mapView)).perform(new BaseViewAction() {
-
-      @Override
-      public void perform(UiController uiController, View view) {
-        GeoJsonSource source = new GeoJsonSource("source", Point.fromLngLat(0d, 0d));
-        mapboxMap.addSource(source);
+    MapboxMapAction.invoke(mapboxMap, (uiController, mapboxMap) -> {
+      GeoJsonSource source = new GeoJsonSource("source", Point.fromLngLat(0d, 0d));
+      mapboxMap.addSource(source);
+      mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
+    });
+  }
 
-        mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
+  @Test
+  public void testFeatureProperties() {
+    validateTestSetup();
+    MapboxMapAction.invoke(mapboxMap, (uiController, mapboxMap) -> {
+      GeoJsonSource source = null;
+      try {
+        source = new GeoJsonSource("source",
+          ResourceUtils.readRawResource(rule.getActivity(), R.raw.test_feature_properties));
+      } catch (IOException exception) {
+        Timber.e(exception);
       }
-
+      mapboxMap.addSource(source);
+      mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
     });
   }
 
   @Test
-  public void testFeatureProperties() throws IOException {
+  public void testUpdateCoalescing() {
     validateTestSetup();
-    onView(withId(R.id.mapView)).perform(new BaseViewAction() {
-
-      @Override
-      public void perform(UiController uiController, View view) {
-        GeoJsonSource source = null;
-        try {
-          source = new GeoJsonSource("source",
-            ResourceUtils.readRawResource(rule.getActivity(), R.raw.test_feature_properties));
-        } catch (IOException exception) {
-          Timber.e(exception);
-        }
-        mapboxMap.addSource(source);
-
-        mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
+    MapboxMapAction.invoke(mapboxMap, (uiController, mapboxMap) -> {
+      GeoJsonSource source = new GeoJsonSource("source");
+      mapboxMap.addSource(source);
+      mapboxMap.addLayer(new CircleLayer("layer", source.getId()));
+
+      source.setGeoJson(Point.fromLngLat(0, 0));
+      source.setGeoJson(Point.fromLngLat(-25, -25));
+      try {
+        source.setGeoJson(ResourceUtils.readRawResource(rule.getActivity(), R.raw.test_feature_properties));
+      } catch (IOException exception) {
+        Timber.e(exception);
       }
 
+      source.setGeoJson(Point.fromLngLat(20, 55));
+      uiController.loopMainThreadForAtLeast(1000);
```

Comment:
1 second timeouts in tests always worry me because eventually the test runs in some (temporarily) really slow environment and fails, and then causes a lot of confusion. Is there a way to explicitly wait for a "map done rendering" event before querying? If not, it's not too big a deal.

Reply:
I was trying to hook into map callbacks before but that wasn't reliable either, or maybe I just wasn't looking at the right things at the time. This is, unfortunately, an approach we've taken with most of the tests that need to wait for the renderer and it's definitely worth exploring for alternatives.